### PR TITLE
Fixed #34714 -- Added aget_object_or_404()/aget_list_or_404() shortcuts.

### PR DESCRIPTION
**Pull request by olibook**
Implementation for https://code.djangoproject.com/ticket/34714

I'm a little worried about code duplication between async and non async versions of both functions. However it seems a common pattern to duplicate 'some' of the code between async and sync functions, eg:

- `acheck_password` /  `check_password` in [base_user.py](https://github.com/django/django/blob/e4a5527d1dc2f8183883931560f3a6dcdef0ab0c/django/contrib/auth/base_user.py#L126)
- `get_many()` / `a_getmany()` in [cache/backends/base.py](https://github.com/django/django/blob/e4a5527d1dc2f8183883931560f3a6dcdef0ab0c/django/core/cache/backends/base.py#L208)

I tried factoring the code with private helper functions without great satisfaction regarding readability.

### CODE REVIEW

File: django/shortcuts.py [hunk 1]
```diff
@@ -114,6 +122,11 @@ def get_list_or_404(klass, *args, **kwargs):
     return obj_list
 
 
+async def aget_list_or_404(klass, *args, **kwargs):
+    """See get_list_or_404()."""
+    return await sync_to_async(get_list_or_404)(klass, *args, **kwargs)
```
Comment (felixxm) [inline django/shortcuts.py:127]: `QuerySet.aget()` and `afilter()` are already implemented. We should provide an async variants not only wrappers on existing functions.

Comment (olibook) [reply django/shortcuts.py:127]: let's go for it !

Comment (olibook) [reply django/shortcuts.py]: Hi @felixxm 

there's no `filter()` for a [very good reason ](https://code.djangoproject.com/ticket/34147). I get the idea.

File: django/shortcuts.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 of MVC. In other words, these functions/classes introduce controlled coupling
 for convenience's sake.
 """
+
```
Comment (felixxm) [inline django/shortcuts.py:6]: Remove unrelated changes.

File: django/shortcuts.py [hunk 3]
```diff
@@ -89,6 +90,25 @@ def get_object_or_404(klass, *args, **kwargs):
...
+async def aget_object_or_404(klass, *args, **kwargs):
+    """See get_object_or_404()."""
+    queryset = _get_queryset(klass)
+    if not hasattr(queryset, "aget"):
+        klass__name = (
+            klass.__name__ if isinstance(klass, type) else klass.__class__.__name__
+        )
+        raise ValueError(
+            "First argument to aget_object_or_404() must be a Model, Manager, "
+            "or QuerySet, not '%s'." % klass__name
```
Comment (felixxm) [inline django/shortcuts.py:102]: I'd use f-strings in new functions.

File: docs/releases/5.0.txt [hunk 4]
```diff
@@ -240,6 +240,12 @@ Asynchronous views
   perform any necessary cleanup if a client disconnects before the response is
   generated. See :ref:`async-handling-disconnect` for more details.
 
+* The following asynchronous shortcuts are implemented:
+
+  * :func:`~django.shortcuts.aget_object_or_404`
+  * :func:`~django.shortcuts.aget_list_or_404`
+
+
```
Comment (felixxm) [inline docs/releases/5.0.txt:248]: Chop blank line.
```suggestion

```

File: docs/releases/5.0.txt [hunk 5]
```diff
@@ -240,6 +240,12 @@ Asynchronous views
   perform any necessary cleanup if a client disconnects before the response is
   generated. See :ref:`async-handling-disconnect` for more details.
 
+* The following asynchronous shortcuts are implemented:
+
+  * :func:`~django.shortcuts.aget_object_or_404`
+  * :func:`~django.shortcuts.aget_list_or_404`
```
Comment (felixxm) [inline docs/releases/5.0.txt:246]: Maybe:
```suggestion
* The new :func:`~django.shortcuts.aget_object_or_404` and
  :func:`~django.shortcuts.aget_list_or_404` asynchronous shortcuts allow
  getting objects from an asynchronous context.
```

**Review by felixxm — COMMENTED** [review]
@olibook Thanks :+1: I left small suggestions.

File: docs/releases/5.0.txt (hunk 5 above)
Comment (olibook) [reply docs/releases/5.0.txt:246]: Yes that's better

File: django/shortcuts.py (hunk 3 above)
Comment (olibook) [reply django/shortcuts.py:102]: good to know

**Comment by olibook** [comment]
Hi @felixxm thanks for the review. I tidied up this PR with your comments.

File: tests/get_object_or_404/tests.py [hunk 6]
```diff
@@ -85,6 +90,88 @@ def test_get_object_or_404(self):
...
+        )
+        self.assertEqual(
+            await aget_list_or_404(
+                Article,
+                Q(title__startswith="Run") | Q(title__startswith="Walk"),
+                authors__name="Patsy",
+            ),
+            [article],
+        )
+
```
Comment (felixxm) [inline tests/get_object_or_404/tests.py]: I moved tests to the `tests/async/test_async_shortcuts.py`.

**Review by felixxm — COMMENTED** [review]
@olibook Thanks :+1:
